### ISSUE TITLE
Primary role deactivation timeout

### ISSUE THREAD

**Issue by jwr** [UltimateHackingKeyboard/agent#792]
The primary/secondary roles for keys are a great idea. However, I find that I often press a modifier key (that has two roles), hold it for a while, and then change my mind. The primary role is then triggered.

I think it would be great to be able to specify a timeout, after which the primary role is no longer triggered. I'm guessing it should be pretty short. The idea is to disable the primary role if the key was held, rather then pressed. I don't know what a good definition of "held" is: 100ms?

The outcome I'd expect would be for me to be able to hold my "Mod" key (which is Space/Mod right now) and release it without getting a space.

**Comment by mondalaci** [comment]
This is a duplicate of #679. Usually, I close the new version of the duplicate in favor of their old version, but this time, I rather close the old issue because this one is more clearly phrased.

This issue is blocked by #562.

**Comment by anilanar** [comment]
It seems this ticket is not blocked anymore :partying_face:

**Comment by mondalaci** [comment]
It's still blocked by #562.

**Comment by kareltucek** [comment]
Btw, multiple possibly-desirable on-timeout behaviours exist:

- No timeout. I.e., current behaviour - wait until something happens.
- Suppress primary action, (but allow secondary even after timeout). (i.e., the behaviour proposed in this ticket)
- Activate primary action. This makes sense on spacebar or enter, because there it makes sense to hold them for longer time (e.g. if I want to get a long row of spaces).
- Activate secondary role (immediately). This is somewhat similar to suppress strategy, but it allows to use secondary role modifiers with external mouse (e.g., shift + click).

Also, lets note that this ticket is loosely connected to the topic of alphabetic-key-friendly resolution strategy - this will need to be discussed further when the time to actually change the config format comes. (Firmware-wise, I can implement either of them or both any time, but config format migration needs to be resolved for either of them. Anyway, I assume that this is in fact more blocked by ongoing development of modules which has much higher priority.)

**Comment by mondalaci** [comment]
The user configuration will be extended by a uint16 `primaryRoleDeactivationTimeout` value measured in milliseconds. When this value is 0, the "no timeout" behavior should apply. Otherwise, the "suppress primary action" behavior should apply.

On second thought, I'm considering to repurpose `doubleTapSwitchLayerTimeout` in the user configuration as `primaryRoleDeactivationTimeout` given that both `doubleTapSwitchLayerTimeout` and `DoubleTapSwitchLayerReleaseTimeout` have been hardcoded in the firmware for ages, and nobody has complained.

Based on how well the above hardcoded values work, I'm not even sure if the exact primary role deactivation timeout value should be specifiable. Maybe using a sane default and simply providing an on/off switch would work well.

**Comment by bodograumann** [comment]
This would also be very useful to use modifiers like Shift, Ctrl or Win as secondary function with the mouse.
They are not only used to modify other keys, thus activating the secondary function, but also together with a mouse click / drag. When using UHK’s mouse mode, this works well-enough, but with a dedicated pointer device, the UHK does not recognize the mouse click and thus still activates the primary function on the modifier after it is eventually released.

**Comment by jwstyle75** [comment]
I set secondary role as LShift on 'S' key.

It works great only when I hold S key as a LShift(I have left pinky finger pain and this seems good solution for me).

but if I press(not hold) S key repeat fastly, sometimes UHK drop some key inputs.

of couse, I press S key repeat slowly, It's OK. but it's not good. it makes typing speed too slow.

this issue is not yet solved i think(this issue up since in Sep, 2018).

Is there any other way to use Secondary Role without  key press drop?

**Comment by kareltucek** [comment]
@jwstyle75 I don't see any connection to this issue, so please let's continue in https://github.com/UltimateHackingKeyboard/firmware/issues/447.

**Comment by mondalaci** [comment]
@kareltucek Does your fork contain a feature like this? If so, we could expose a relevant smart macro variable.

**Comment by kareltucek** [comment]
No. Not in the sense of a parametric property of the resolution strategy.

**Comment by mondalaci** [comment]
How hard would it be to implement it? I think it'd be useful to have such a setting. Feel free to let me know if you disagree or have a better idea.

**Comment by kareltucek** [comment]
Not hard. But entangled a lot with [firmware#326 ](https://github.com/UltimateHackingKeyboard/firmware/pull/326) and with tasks that should be done when 326 is merged. (Namely removing the corresponding code from macro engine and reconnecting the `ifPrimary/ifSecondary` commands to work with the refactored secondary role driver.)

**Comment by robbievogt** [comment]
Is there any update on this? It looks like 326 has been abandoned, so I'm assuming no progress?

If it's possible to achieve the desired behaviour through macros, could examples be added here and in the Knowledge Base article pointing here (https://ultimatehackingkeyboard.com/knowledgebase/does-the-uhk-support-dual-role-keys). 

Thanks!

**Comment by kareltucek** [comment]
I plan to port that PR eventually.

**Comment by robbievogt** [comment]
For anyone else following this trail, there are examples of using macros to achieve this functionality in https://github.com/UltimateHackingKeyboard/firmware/blob/v9.2.2/doc-dev/user-guide.md#time-machine.

The example "space cadet shift" was pretty much what I was looking for.

**Comment by mhantsch** [comment]
I think this issue can be closed, as the functionality can now be done through extended macros. Here is an example: hold key for mod (secondary), tap key for backspace (primary), will NOT produce primary backspace if held for more than 300ms.

```
holdLayer mod
ifInterrupted break
ifPlaytime 300 break
tapKey backspace
'''

Store this as a macro (macro action "command"), then bind the key to the macro.

**Comment by mondalaci** [comment]
I think it still makes practical sense to make the timeout adjustable globally.

**Comment by mondalaci** [comment]
Done by https://github.com/UltimateHackingKeyboard/firmware/pull/624
